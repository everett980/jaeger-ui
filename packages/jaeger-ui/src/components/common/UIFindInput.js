// @flow

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { /* Icon, */ Input } from 'antd';
import _debounce from 'lodash/debounce';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import queryString from 'query-string';

import type { Location, RouterHistory } from 'react-router-dom';

import updateUIFind from '../../utils/update-ui-find';

import type { ReduxState } from '../../types/index';

type propsType = {
  inputProps?: Object,
  history: RouterHistory,
  location: Location,
  registerUIFindInputRef?: () => void,
  uiFind?: string,
};

type stateType = {
  ownInputValue: ?string,
};

export class UnconnectedUIFindInput extends React.PureComponent<propsType, stateType> {
  static defaultProps = {
    inputProps: {},
    registerUIFindInputRef: null,
    uiFind: null,
  };

  constructor(props: propsType) {
    super(props);
    this.state = {
      ownInputValue: null,
    };
  }

  handleInputBlur = () => {
    this.updateUIFindQueryParam.flush();
    this.setState({ ownInputValue: null });
  };

  handleInputChange = (evt: SyntheticInputEvent<HTMLInputElement>) => {
    const { value } = evt.target;
    this.updateUIFindQueryParam(value);
    this.setState({ ownInputValue: value });
  };

  updateUIFindQueryParam = _debounce((uiFindQueryParam: ?string) => {
    const { history, location } = this.props;
    const arg = {
      location,
      history,
      uiFind: uiFindQueryParam,
    };
    updateUIFind(arg);
  }, 250);

  render() {
    const inputValue =
      typeof this.state.ownInputValue === 'string' ? this.state.ownInputValue : this.props.uiFind;
    return (
      <Input
        {...this.props.inputProps}
        className="UIFind--input"
        onBlur={this.handleInputBlur}
        onChange={this.handleInputChange}
        ref={this.props.registerUIFindInputRef}
        value={inputValue}
      />
    );
  }
}

export function extractUIFindFromState(state: ReduxState): { uiFind?: string } {
  const { uiFind } = queryString.parse(state.router.location.search);
  return { uiFind };
}

export default withRouter(connect(extractUIFindFromState)(UnconnectedUIFindInput));
