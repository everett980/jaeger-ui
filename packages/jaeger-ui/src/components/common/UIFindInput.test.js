// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { shallow } from 'enzyme';
import { Input } from 'antd';
import debounceMock from 'lodash/debounce';
import queryString from 'query-string';

import { UnconnectedUIFindInput, extractUIFindFromState } from './UIFindInput';
import updateUIFindSpy from '../../utils/update-ui-find';

jest.mock('lodash/debounce');

jest.mock('../../utils/update-ui-find');

describe('UIFind', () => {
  const flushMock = jest.fn();
  const queryStringParseSpy = jest.spyOn(queryString, 'parse');

  const uiFind = 'uiFind';
  const ownInputValue = 'ownInputValue';
  const props = {
    uiFind: null,
    history: {
      replace: () => {},
    },
    location: {
      search: null,
    },
  };
  let wrapper;

  beforeAll(() => {
    debounceMock.mockImplementation(fn => {
      function debounceFunction(...args) {
        fn(...args);
      }
      debounceFunction.flush = flushMock;
      return debounceFunction;
    });
  });

  beforeEach(() => {
    flushMock.mockReset();
    wrapper = shallow(<UnconnectedUIFindInput {...props} />);
  });

  describe('rendering', () => {
    it('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders props.uiFind when state.ownInputValue is `null`', () => {
      wrapper.setProps({ uiFind });
      expect(wrapper.find(Input).prop('value')).toBe(uiFind);
    });

    it('renders state.ownInputValue when it is not `null` regardless of props.uiFind', () => {
      wrapper.setProps({ uiFind });
      wrapper.setState({ ownInputValue });
      expect(wrapper.find(Input).prop('value')).toBe(ownInputValue);
    });
  });

  describe('typing in input', () => {
    const newValue = 'newValue';

    it('updates state', () => {
      wrapper.find(Input).simulate('change', { target: { value: newValue } });
      expect(wrapper.state('ownInputValue')).toBe(newValue);
    });

    it('calls updateUIFind with correct kwargs', () => {
      wrapper.find(Input).simulate('change', { target: { value: newValue } });
      expect(updateUIFindSpy).toHaveBeenLastCalledWith({
        history: props.history,
        location: props.location,
        uiFind: newValue,
      });
    });
  });

  describe('blurring input', () => {
    it('clears state.ownInputValue', () => {
      wrapper.setState({ ownInputValue });
      expect(wrapper.state('ownInputValue')).toBe(ownInputValue);
      wrapper.find(Input).simulate('blur');
      expect(wrapper.state('ownInputValue')).toBe(null);
    });

    it('triggers pending queryParameter updates', () => {
      wrapper.find(Input).simulate('blur');
      expect(flushMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('extractUIFindFromState', () => {
    beforeAll(() => {
      queryStringParseSpy.mockReturnValue({ uiFind });
    });

    it('returns uiFind from parsed state.router.location.search', () => {
      const reduxStateValue = 'state.router.location.search';
      const result = extractUIFindFromState({
        router: {
          location: {
            search: reduxStateValue,
          },
        },
      });
      expect(queryStringParseSpy).toHaveBeenCalledWith(reduxStateValue);
      expect(result).toEqual({
        uiFind,
      });
    });
  });
});
