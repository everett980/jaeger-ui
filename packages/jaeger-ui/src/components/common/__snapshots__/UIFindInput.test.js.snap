// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIFind rendering renders as expected 1`] = `
<Input
  className="UIFind--input"
  disabled={false}
  onBlur={[Function]}
  onChange={[Function]}
  prefixCls="ant-input"
  type="text"
  value={null}
/>
`;
